reference-delegate-explainer.md

```diff
@@ -0,0 +1,562 @@
+  > Note: This is in contrast to the ARIAMixin attributes like `ariaLabelledByElements`, which _do_ unlock the new functionality of referring out of the shadow DOM. In that case, the complexity is worthwhile to include in the ARIAMixin design.
+* At a basic level, Reference Delegate is augmenting the existing functionality of referring to elements by ID string. It seems in line with the design to require using ID strings.
+
+### Use separate attributes for each forwarded attribute
+
+An alternative to a single attribute `shadowrootreferencedelegatemap` / `ShadowRoot.referenceDelegateMap` would be to have individual attributes for each forwarded attribute:
+* `shadowrootariaactivedescendantdelegate`
+* `shadowrootariacontrolsdelegate`
+* `shadowrootariadescribedbydelegate`
+* `shadowrootariadetailsdelegate`
+* `shadowrootariaerrormessagedelegate`
+* `shadowrootariaflowtodelegate`
+* `shadowrootarialabelledbydelegate`
+* `shadowrootariaownsdelegate`
+* `shadowrootfordelegate`
+* `shadowrootformdelegate`
+* `shadowrootlistdelegate`
+* `shadowrootpopovertargetdelegate`
+* `shadowrootinvoketargetdelegate`
+* `shadowrootinteresttargetdelegate`
+* `shadowrootheadersdelegate`
+* `shadowrootitemrefdelegate`
+* `shadowrootreferencedelegate` -- all other references except the ones specified above
+
+Reflected by JavaScript attributes `ShadowRoot.ariaActiveDescendantDelegate`, etc.
+
+#### Pros
+
+* Syntax is more in line with other HTML attributes, rather than using a comma-separated list of colon-separated map entries.
+* Works with IDs that contain commas.
+
+#### Cons
+
+* Adds 15+ new attributes instead of 2.
+* Less clear(?) that `shadowrootreferencedelegate` only forwards references that are not explicitly specified by other elements.
+
+### Designate target elements using attributes instead of IDREF
+
+The [Cross-root ARIA Reflection API](https://github.com/Westbrook/cross-root-aria-reflection/blob/main/cross-root-aria-reflection.md#usage-with-declarative-shadow-dom) explainer proposes adding attributes to elements inside the shadow tree:
+```html
+<x-foo id="foo">
+  <template shadowroot="open" shadowrootreflectscontrols shadowrootreflectsariaactivedescendent>
+    <ul reflectariacontrols>
+      <li>Item 1</li>
+      <li reflectariaactivedescendent>Item 2</li>
+      <li>Item 3</li>
+    </ul>
+  </template>
+</x-foo>
+```
+
+#### Pros
+
+* Does not require an ID on the target element. [But does still require an extra attribute; possibly in addition to an ID if that ID is used for other purposes.]
+
+#### Cons
+
+* Requires new attributes in two places in order to work: E.g. `shadowrootreflectscontrols` on the shadow root _and_ `reflectariacontrols` on the target element.
+* When multiple elements are used for the same attribute, the author cannot control the order (the order is always the DOM order).
+
+### Use exported IDs instead of per-attribute targeting
```

Comment:
I'd like to make one last pitch for this, and then I promise I'll let it go :) 

**The default `referenceDelegate` attribute (without the per-attribute targeting) supports a fairly well-understood and easy to articulate use case: the "semantic delegate" case.**

We've been using this term to refer to composing behaviour for custom elements by using a built-in (or even custom) element inside a shadow root to reuse its functionality (a [best practice for accessibility](https://www.w3.org/TR/using-aria/#rule1) in particular), and layering in extra content and functionality encapsulated within the custom element. The custom element/shadow host then "acts like" the built-in element, but with custom behaviour and/or bundled content. This is a pattern [already widely used](https://github.com/alice/aom/blob/gh-pages/semantic-delegate.md#some-examples-of-custom-elements-from-component-libraries-which-use-this-pattern) in custom element libraries.

`referenceDelegate` supports the "semantic delegate" use case by solving the problem that, conceptually, the custom element _should_ be the target for certain IDREF attributes, but those attributes need to target the "delegate" element in practice. This preserves the encapsulation of the custom element by not requiring authors using the element to know anything about those implementation details. They can simply use the custom element as though it _is_ the delegate element[1].

**I think `referenceDelegateMap` is trying to solve _other problems_**, and I don't think it's solving them anywhere near as well.

The only obvious[2] case I can come up with for `referenceDelegateMap` is the `aria-activedescendant` case, like where `<input>` uses `aria-activedescendant` to refer to one of the `option`s inside the shadow root of a custom element representing a `listbox` in the [Combining `referenceDelegate` and `referenceDelegateMap`](https://github.com/behowell/aom/blob/reference-delegate/reference-delegate-explainer.md#combining-referencedelegate-and-referencedelegatemap) example. 

In this case, you _must know_ that there are `option` elements within the `<fancy-listbox>` which can be a valid target for `aria-activedescendant`, and you also _must know_ that the attribute will be forwarded correctly. This is a completely different case from the semantic delegate case, in which you _don't need to know anything about the internals_, only that you can treat the host as though it is what it "appears" to be.

**To me, `exportid` is a better fit for this type of problem**: where there is some element inside the shadow root that needs to be targeted explicitly, but for whatever reason needs to remain inside the shadow root. `exportid` allows you to specify an "API" for your element to allow references to elements inside its shadow root that you guarantee will exist, but need not reveal any more information about.

**`exportid` also gives us ample flexibility for unforeseen cases.** What I would especially like to avoid is a situation where we figure out all of the inherent complexity of supporting `referenceDelegateMap` (e.g. how to design its JavaScript API) and then realise that there actually are cases where the "bottleneck problem" is an issue, so we need to do something like `exportid` anyway. Not only would we then have spent extra effort in designing `referenceDelegateMap`, but now authors have to choose which one to use for their largely overlapping use cases.

---

[1] Leaving aside that you can't use IDREF attributes on the host and also have them forwarded, which was deliberately left out of scope here to keep the scope as minimal as possible.

[2] I would really like to understand what other cases there are! @nolanlawson you mentioned doing an audit of your component library recently; perhaps you have some examples? The example under [Delegating to multiple elements](https://github.com/behowell/aom/blob/reference-delegate/reference-delegate-explainer.md#delegating-to-multiple-elements) is interesting, but seems like it would equally be a problem in a non-shadow root world. That is, assuming the content inside the shadow root in this example is intended to be "unknown" (i.e. user-generated or similar) to the page author, the author would still have no way in practice to correctly set `aria-describedby` on the `<input>` even if that content were in light DOM. And if it was user-generated, how would you correctly set the `referenceDelegateMap`? But, maybe this is all wishful thinking on my part :)

Reply:
@alice The `ariaActiveDescendant` case is definitely the strongest one I've seen. The fact that you want an `ariaControls` as well as an `ariaActiveDescendant` pointing into the same shadow root came up twice in my component audit (both were comboboxes).

Aside from that, every use case I've seen for any kind of cross-root references could be solved by either a "semantic delegate" type solution, or [element reflection](https://chromestatus.com/feature/6244885579431936) rather than reference delegates – including the tricky case of multiple elements inside one shadow root sharing the same ARIA attribute (see our [lightning-input](https://developer.salesforce.com/docs/component-library/bundle/lightning-input/specification) and its `date-aria-*` and `time-aria-*` props, which allow for a separate date element and time element to each point to separate elements in their shadow-including-ancestor using `controls`/`describedBy`/`labelledBy`). This has led me to believe that the "bottleneck effect" is not such a dealbreaker for the current proposal.

I definitely agree though that it feels "weird" to have (say) the listbox be both the target for `controls` as well as `activeDescendant`. But I would still be happy with this proposal since it gives us a path forward – perhaps userland solutions can make it more ergonomically appealing. 🙂